Assets/Scripts/Communication/CommunicationManager.cs

```diff
@@ -27,120 +30,136 @@ public class CommunicationManager
+		_responseManager = new Thread(processResponses);
+		_responseManager.Start();
 		Debug.Log("Communication Manager has started successfully");
 	}
 
-	static void ProcessRequests(){
-		_threadRunning = true;
-		Dictionary<string, object> request = null;
+	static void processResponses(){
+		_responseThreadRunning = true;
 		Dictionary<string, object> response = null;
-		bool done = false;
 
 		try{
-			while (_threadRunning){
-				request = null;
+			while (_responseThreadRunning){
 				response = null;
-
-				//Take a request off the queue
-				lock(requestQueueLock){
-					if (requestQueue.Count > 0){
-						request = requestQueue.Dequeue();
-					}
-				}
-
-				string requestID = "";
-				if (request != null){
-					//Send the request to the server
-					requestID = (string)request["Request_ID"];
-					Debug.Log("Manager Thread Sending Request: " + requestID);
-					SendCommand(request);
-				}
 				
-				done = false;
-				while (!done){
-					Dictionary<string, object> resp = null;
+				lock (responseDictLock){
+					//Suspend the thread until the websocket receives a message
+					Monitor.Wait(responseDictLock);
+
 					try{
-						resp = GetNextResponse();
+						response = GetNextResponse();
 					} catch (Exception e){
-						if (e.ToString().Contains("unauthenticated")){
-							if (request != null){
-								SendCommand(request);
-							}
-						}
 						Debug.Log("Exception getting response: " + e.ToString());
-						throw e;
-					}
-					
-					if (resp == null){
-						break;
 					}
 
-					if (resp.ContainsKey("Request_ID") && (string) resp["Request_ID"] == requestID){
-						response = resp;
-					}
-					else{
-						lock(asyncMessagesLock){
-							Debug.Log("Found an async message");
-							LogDictionary(resp);
-							asyncMessagesQueue.Enqueue(resp);
-							Debug.Log("Async Queue Size: " + asyncMessagesQueue.Count.ToString());
+					if (response != null){
+						//Add the response to the common dictionaries
+						if (response.ContainsKey("Request_ID")){
+							string requestID = (string) response["Request_ID"];
+							Debug.Log("Obtained response for request " + requestID);
+							responseDict[requestID] = response;
+						}
+						else{
+							Debug.Log("Obtained async message from server");
+							lock(asyncMessagesLock){
+								asyncMessagesQueue.Enqueue(response);
+							}
 						}
 					}
 				}
 				
-				if (response != null){
-					//Add the response to the common dictionary
-					lock(responseDictLock){
-						responseDict[requestID] = response;
+			}
+		} catch (Exception e){
+			Debug.Log("Response Manager Thread crashed: " + e.ToString());
+		}
+		
+	}
+
+	static void ProcessRequests(){
+		_requestThreadRunning = true;
+		Dictionary<string, object> request = null;
+
+		try{
+			while (_requestThreadRunning){
+				request = null;
+
+				//Take a request off the queue
+				lock(requestQueueLock){
+					//Suspend the thread until the websocket receives a message
+					Monitor.Wait(requestQueueLock);
```

Comment:
I thought the same thing at first but this is actually how you are supposed to do it.  If you don't put it in a lock it throws an exception.  It appears that the .wait gives up the lock.